klayout/drc/run_drc.py

```diff
@@ -0,0 +1,709 @@
+    --verbose                           Detailed rule execution log for debugging.
+"""
+
+from docopt import docopt
+import os
+import xml.etree.ElementTree as ET
+import logging
+import klayout.db
+import glob
+from datetime import datetime
+from subprocess import check_call
+import shutil
+import concurrent.futures
+import traceback
+
+
+def get_rules_with_violations(results_database):
+    """
+    This function will find all the rules that has violated in a database.
+
+    Parameters
+    ----------
+    results_database : string or Path object
+        Path string to the results file
+
+    Returns
+    -------
+    set
+        A set that contains all rules in the database with violations
+    """
+
+    mytree = ET.parse(results_database)
+    myroot = mytree.getroot()
+
+    all_violating_rules = set()
+
+    for z in myroot[7]:  # myroot[7] : List rules with viloations
+        all_violating_rules.add(f"{z[1].text}".replace("'", ""))
+
+    return all_violating_rules
+
+
+def check_drc_results(results_db_files: list):
+    """
+    check_drc_results Checks the results db generated from run and report at the end if the DRC run failed or passed.
+    This function will exit with 1 if there are violations.
+
+    Parameters
+    ----------
+    results_db_files : list
+        A list of strings that represent paths to results databases of all the DRC runs.
+    """
+
+    if len(results_db_files) < 1:
+        logging.error("Klayout did not generate any rdb results. Please check run logs")
+        exit(1)
+
+    full_violating_rules = set()
+
+    for f in results_db_files:
+        violating_rules = get_rules_with_violations(f)
+        full_violating_rules.update(violating_rules)
+
+    if len(full_violating_rules) > 0:
+        logging.error("Klayout DRC run is not clean.")
+        logging.error("Violated rules are : {}\n".format(str(full_violating_rules)))
+        exit(1)
+    else:
+        logging.info("Klayout DRC run is clean. GDS has no DRC violations.")
+
+
+def get_results(rule_deck, rules, lyrdb, type):
+
+    mytree = ET.parse(f"{lyrdb}_{type}_gf{arguments['--gf180mcu']}.lyrdb")
+    myroot = mytree.getroot()
+
+    violated = []
+
+    for lrule in rules:
+        # Loop on database to get the violations of required rule
+        for z in myroot[7]:
+            if f"'{lrule}'" == f"{z[1].text}":
+                violated.append(lrule)
+                break
+
+    lyrdb_clean = lyrdb.split("/")[-1]
+
+    if len(violated) > 0:
+        logging.error(
+            f"\nTotal # of DRC violations in {rule_deck}.drc is {len(violated)}. Please check {lyrdb_clean}_{type}_gf{arguments['--gf180mcu']}.lyrdb file For more details"
+        )
+        logging.info("Klayout GDS DRC Not Clean")
+        logging.info(f"Violated rules are : {violated}\n")
+    else:
+        logging.info(
+            f"\nCongratulations !!. No DRC Violations found in {lyrdb_clean} for {rule_deck}.drc rule deck with switch gf{arguments['--gf180mcu']}"
```

Comment:
usually it's good to keep that kind of user facing logging close to the top level of the codebase and use structured return values to convey the status back to the caller.

Reply:
removed